### PR TITLE
[feature][processing] A brand new spatiotemporal ST-DBSCAN clustering algorithm

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR implements a brand new spatiotemporal ST-DBSCAN clustering algorithm in QGIS' processing toolbox.

It's a pretty straightforward copy/paste of @nyalldawson 's DBSCAN clustering algorithm, with a temporal component added.

Obligatory screenshot:
image

### CODE REVIEW

**Comment by nirvn** [comment]
@anitagraser , that's something that might interest you.

File: src/analysis/processing/qgsalgorithmstdbscanclustering.cpp [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Clusters" ), QgsProcessing::TypeVectorPoint ) );
+
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "NUM_CLUSTERS" ), QObject::tr( "Number of clusters" ) ) );
+}
+
+QString QgsStDbscanClusteringAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Clusters point features based on a 2D implementation of Density-based spatiotemporal clustering of applications with noise (ST-DBSCAN) algorithm.\n\n"
+                      "The algorithm requires three parameters, a minimum cluster size (“minPts”), the maximum distance allowed between clustered points (“eps1”), and the maximum datetime interval allowed between cluster points (“eps2”)." );
```
Comment (DelazJ) [inline src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:91]: Is this necessary? It simply mentions what is shown in the GUI. I'd instead expect explanation on what/how the "Density-based spatiotemporal clustering ..." is/works.
Also, it leads to the question: does it mean that the date/time field is/should be (tagged) optional?

Comment (nirvn) [reply src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:91]: @DelazJ , the help string matches what we have for the DBSCAN clustering algorithm.

And no, the date/time field is _not_ optional, it'd defeat the whole purpose :) for a spatial-only clustering, DBSCAN is the algorithm to use.

Comment (DelazJ) [reply src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:91]: > the help string matches what we have for the DBSCAN clustering algorithm.

Maybe the DBSCAN help as is is not the best template...?
> And no, the date/time field is not optional, it'd defeat the whole purpose :)

Agreed 😄. What I meant is: it's stated **`requires three parameters`** and there are 4 in the GUI (with none optional)

Comment (nirvn) [reply src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:91]: I'll just go ahead and remove the paragraph then.

File: src/analysis/processing/qgsalgorithmstdbscanclustering.cpp [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+}
+
+QString QgsStDbscanClusteringAlgorithm::displayName() const
+{
+  return QObject::tr( "ST-DBSCAN clustering" );
+}
+
+QString QgsStDbscanClusteringAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Clusters spatiotemporal point features using a density based scan algorithm." );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:36]:
```suggestion
  return QObject::tr( "Clusters spatiotemporal point features using a time and density based scan algorithm." );
```
?

File: src/analysis/processing/qgsalgorithmstdbscanclustering.cpp [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+}
+
+QStringList QgsStDbscanClusteringAlgorithm::tags() const
+{
+  return QObject::tr( "clustering,clusters,density,based,points,temporal,time" ).split( ',' );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:41]: maybe add "interval" and "distance" here too?

File: src/analysis/processing/qgsalgorithmstdbscanclustering.cpp [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+  addParameter( new QgsProcessingParameterField( QStringLiteral( "DATETIME_FIELD" ),
+                QObject::tr( "Date/time field" ), QVariant(),
+                QStringLiteral( "INPUT" ), QgsProcessingParameterField::DateTime, false, false ) );
+
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "MIN_SIZE" ), QObject::tr( "Minimum cluster size" ),
+                QgsProcessingParameterNumber::Integer, 5, false, 1 ) );
+  addParameter( new QgsProcessingParameterDistance( QStringLiteral( "EPS1" ),
+                QObject::tr( "Maximum distance between clustered points" ), 1, QStringLiteral( "INPUT" ), false, 0 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "EPS2" ),
+                QObject::tr( "Maximum date/time interval (in days unit) between clustered points" ), QgsProcessingParameterNumber::Double, 1, false, 0 ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:67]: can you explain the rationale behind using "days" as a unit, vs say hours?

File: src/analysis/processing/qgsalgorithmstdbscanclustering.cpp [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+  addParameter( sizeFieldNameParam.release() );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Clusters" ), QgsProcessing::TypeVectorPoint ) );
+
+  addOutput( new QgsProcessingOutputNumber( QStringLiteral( "NUM_CLUSTERS" ), QObject::tr( "Number of clusters" ) ) );
+}
+
+QString QgsStDbscanClusteringAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "Clusters point features based on a 2D implementation of spatiotemporal density-based clustering of applications with noise (ST-DBSCAN) algorithm." );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:90]: Afaics this algorithm is based on https://www.sciencedirect.com/science/article/pii/S0169023X06000218 -- correct? If so, would be great to reference that paper in the short help string!

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:90]: Or even better, reference all the papers from the readme here: https://github.com/eren-ck/st_dbscan

**Review by nyalldawson — COMMENTED** [review]
The algorithm looks good! I'm concerned that there's a heck of a lot of almost identical code shared between this algorithm and QgsDbscanClusteringAlgorithm... I think the two algorithms could easily share a common base to avoid this!

File: src/analysis/processing/qgsalgorithmstdbscanclustering.cpp (hunk 4 above)
Comment (nirvn) [reply src/analysis/processing/qgsalgorithmstdbscanclustering.cpp:65]: Well, the short term rationale is that I felt days as a unit was a good fixed-unit type sitting in between lower units (hours, minutes) and higher units (weeks, months, years).

But it's also a placeholder until I have some free time to code a QgsProcessingParameterInterval, which will have a dropdown combobox to pick the unit type. 

Until then, I think days is tolerable :wink:

**Comment by nirvn** [comment]
@nyalldawson , alright, comments addressed and code de-duplicated.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**